File: money_exchange_rate/src/main/java/com/tools/money/ExchangeRateFragment.java
package com.tools.money;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;

import com.tools.money.injection.DaggerMoneyExchangeComponent;
import com.tools.money_exchange_rate.R;

import javax.inject.Inject;


public class ExchangeRateFragment extends Fragment implements AdapterView.OnItemSelectedListener, View.OnClickListener {

    ExchangeRateViewModel mViewModel;

    @Inject
    ViewModelFactory viewModelFactory;

    private Spinner spFromCurrency;
    private Spinner spToCurrency;
    private Button btConvert;
    private EditText etAmount;
    private String fromCurrency;
    private String toCurrency;
    private TextView tvConvertRate;

    public static ExchangeRateFragment newInstance() {
        return new ExchangeRateFragment();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        DaggerMoneyExchangeComponent.create().inject(this);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mViewModel = ViewModelProviders.of(this, viewModelFactory).get(ExchangeRateViewModel.class);
        getCurrencies();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.exchange_rate_fragment, container, false);
        initView(view);
        return view;
    }

    private void initView(View view) {
        spFromCurrency = view.findViewById(R.id.spinner_from_currency);
        spToCurrency = view.findViewById(R.id.spinner_to_currency);
        btConvert = view.findViewById(R.id.bt_convert);
        etAmount = view.findViewById(R.id.et_amount);
        tvConvertRate = view.findViewById(R.id.tv_convert_rate);

        spFromCurrency.setOnItemSelectedListener(this);
        spToCurrency.setOnItemSelectedListener(this);

        btConvert.setOnClickListener(this);
    }

    private void getCurrencies() {
        //  Stream<String> streamEmpty = Stream.empty();


        mViewModel.getCurrencies().observe(this, currencies -> {
            ArrayAdapter<String> currencyAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, currencies);
            currencyAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            spFromCurrency.setAdapter(currencyAdapter);
            spToCurrency.setAdapter(currencyAdapter);
        });
    }


    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        if (parent.getId() == R.id.spinner_from_currency) {
            fromCurrency = (String) parent.getItemAtPosition(position);
        } else {
            toCurrency = (String) parent.getItemAtPosition(position);
        }

    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    @Override
    public void onClick(View v) {
        mViewModel.getConvertRate(fromCurrency, toCurrency, etAmount.getText().toString()).observe(this, result -> {
            tvConvertRate.setText(result);
        });
    }
}

File: money_exchange_rate/src/main/java/com/tools/money/ExchangeRateFragment.kt
package com.tools.money

import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.Spinner
import androidx.appcompat.content.res.AppCompatResources
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import com.tools.core.BaseFragment
import com.tools.core.network.Status
import com.tools.jackofall.di.FeatureModuleDependency
import com.tools.money.injection.DaggerMoneyExchangeComponent
import com.tools.money_exchange_rate.R
import com.tools.money_exchange_rate.databinding.ExchangeRateFragmentBinding
import dagger.hilt.android.EntryPointAccessors
import timber.log.Timber
import javax.inject.Inject

class ExchangeRateFragment : BaseFragment<ExchangeRateFragmentBinding>(), OnItemSelectedListener {

    @Inject
    lateinit var viewModelFactory: ViewModelFactory


    val exchangeRateViewModel: ExchangeRateViewModel by viewModels { viewModelFactory }

    private lateinit var spFromCurrency: Spinner
    private lateinit var spToCurrency: Spinner
    override fun onAttach(context: Context) {
        super.onAttach(context)
        DaggerMoneyExchangeComponent.builder()
                .appDependencies(
                        EntryPointAccessors.fromApplication(
                                context,
                                FeatureModuleDependency::class.java
                        )
                )
                .build()
                .inject(this)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        //  viewModel = ViewModelProviders.of(this, viewModelFactory).get(ExchangeRateViewModel::class.java)
        binding.viewModel = exchangeRateViewModel
        binding.apply {
            lifecycleOwner = viewLifecycleOwner
        }
        initView()
        getCurrencies()
    }

    override fun onResume() {
        super.onResume()
        setToolBar()
    }

    private fun setToolBar() {
        setTitle("Money Exchange")
        val icon = AppCompatResources.getDrawable(requireContext(), com.tools.jackofall.R.drawable.ic_back)
        icon?.let(::setNavigationIcon)
    }


    private fun initView() {
        spFromCurrency = binding.spinnerFromCurrency
        spToCurrency = binding.spinnerToCurrency
        spFromCurrency.onItemSelectedListener = this
        spToCurrency.onItemSelectedListener = this
    }

    private fun getCurrencies() {
        exchangeRateViewModel.currenciesLiveData.observe(viewLifecycleOwner, Observer {

            it?.let { resource ->
                when (resource.status) {
                    Status.SUCCESS -> {
                        val currencyAdapter = ArrayAdapter<String>(requireContext(), android.R.layout.simple_spinner_item, resource.data?.currencies?.keys?.toList()
                                ?: emptyList())
                        currencyAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                        spFromCurrency.adapter = currencyAdapter
                        spToCurrency.adapter = currencyAdapter
                        val historyFromCurrency = currencyAdapter.getPosition(exchangeRateViewModel.fromCurrency)
                        val historyToCurrency = currencyAdapter.getPosition(exchangeRateViewModel.toCurrency)
                        spFromCurrency.setSelection(historyFromCurrency)
                        spToCurrency.setSelection(historyToCurrency)
                    }
                    Status.ERROR -> {
                        Timber.d("Error while getting currency data")
                        showErrorAlertMessage()
                    }

                }
            }
        })

    }


    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        if (parent.id == R.id.spinner_from_currency) {
            exchangeRateViewModel.fromCurrency = parent.getItemAtPosition(position) as String
        } else {
            exchangeRateViewModel.toCurrency = parent.getItemAtPosition(position) as String
        }
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}


    companion object {
        @JvmStatic
        fun newInstance(): ExchangeRateFragment {
            return ExchangeRateFragment()
        }
    }

    override fun getFragmentLayout() = R.layout.exchange_rate_fragment

    override fun getToolBar() = binding.appbar.toolbar


}